hotshot-task-impls/src/helpers.rs

```diff
@@ -996,3 +1005,107 @@ pub async fn validate_qc_and_next_epoch_qc<TYPES: NodeType, V: Versions>(
+                && Some(second_vid_share.data.payload_commitment())
+                    == da_cert.data().next_epoch_payload_commit)
+        {
+            return Ok(second_vid_share);
+        }
+    }
+
+    let wait_duration = Duration::from_millis(timeout / 2);
+
+    // TODO configure timeout
+    let Some(time_spent) = Instant::now().checked_duration_since(view_start_time) else {
+        // Shouldn't be possible, now must be after the start
+        return Err(warn!(
+            "Now is earlier than the view start time. Shouldn't be possible."
+        ));
+    };
+    let Some(time_left) = wait_duration.checked_sub(time_spent) else {
+        // No time left
+        return Err(warn!("Run out of time waiting for the second VID share."));
+    };
+    let receiver = receiver.clone();
+    let Ok(Some(event)) = tokio::time::timeout(time_left, async move {
+        let da_cert_clone = da_cert.clone();
+        EventDependency::new(
+            receiver,
+            Box::new(move |event| {
+                let event = event.as_ref();
+                if let HotShotEvent::VidShareValidated(second_vid_share) = event {
+                    if target_epoch == da_cert_clone.epoch() {
+                        second_vid_share.data.payload_commitment()
+                            == da_cert_clone.data().payload_commit
+                    } else {
+                        Some(second_vid_share.data.payload_commitment())
+                            == da_cert_clone.data().next_epoch_payload_commit
+                    }
+                } else {
+                    false
+                }
+            }),
+        )
+        .completed()
+        .await
+    })
+    .await
+    else {
+        // Check again, there is a chance we missed it
+        let maybe_second_vid_share = consensus
+            .read()
+            .await
+            .vid_shares()
+            .get(&vid_share.data.view_number())
+            .and_then(|key_map| key_map.get(vid_share.data.recipient_key()))
+            .and_then(|epoch_map| epoch_map.get(&target_epoch))
+            .cloned();
+        if let Some(second_vid_share) = maybe_second_vid_share {
+            if (target_epoch == da_cert.epoch()
+                && second_vid_share.data.payload_commitment() == da_cert.data().payload_commit)
+                || (target_epoch != da_cert.epoch()
+                    && Some(second_vid_share.data.payload_commitment())
+                        == da_cert.data().next_epoch_payload_commit)
+            {
+                return Ok(second_vid_share);
+            }
+        }
+        return Err(warn!("Error while waiting for the second VID share."));
+    };
+    let HotShotEvent::VidShareValidated(second_vid_share) = event.as_ref() else {
+        // this shouldn't happen
+        return Err(warn!("Received event is not VidShareValidated but we checked it earlier. Shouldn't be possible."));
+    };
+    Ok(second_vid_share.clone())
```

Comment:
This kind of logic is why `EventDependency` exists.  Checking shared state over and over is not a good pattern.  I think we should rewrite this to just wait for a VidShareValidated event where the epoch number matches what we want